tests/50federation/36-state.pl

```diff
@@ -46,3 +46,150 @@
+   requires => [ $main::OUTBOUND_CLIENT, $main::INBOUND_SERVER, $main::HOMESERVER_INFO[0],
+                 local_user_and_room_fixtures(),
+                 federation_user_id_fixture() ],
+
+   do => sub {
+      my ( $outbound_client, $inbound_server, $info, $creator, $room_id, $user_id ) = @_;
+      my $first_home_server = $info->server_name;
+
+      my $local_server_name = $outbound_client->server_name;
+
+      my $room;
+      my $sent_event;
+
+      $outbound_client->join_room(
+         server_name => $first_home_server,
+         room_id     => $room_id,
+         user_id     => $user_id,
+      )->then( sub {
+         ( $room ) = @_;
+
+         # Generate but don't send an event
+         my $missing_event = $room->create_event(
+            type => "m.room.message",
+
+            sender  => $user_id,
+            content => {
+               body => "Message 1",
+            },
+         );
+
+         # Generate another one and do send it so it will refer to the
+         # previous in its prev_events field
+         $sent_event = $room->create_event(
+            type => "m.room.message",
+
+            # This would be done by $room->create_event anyway but lets be
+            #   sure for this test
+            prev_events => [
+               [ $missing_event->{event_id}, $missing_event->{hashes} ],
+            ],
+
+            sender  => $user_id,
+            content => {
+               body => "Message 2",
+            },
+         );
+
+         Future->needs_all(
+            $inbound_server->await_request_state_ids( $room_id )
+            ->then( sub {
+               my ( $req ) = @_;
+
+               log_if_fail "Got /state_ids";
+
+               $req->respond( HTTP::Response->new( 404, "Not Found", [ Content_Length => 0 ] ) );
+
+               Future->done(1);
+            }),
+            $inbound_server->await_request_state( $room_id )
+            ->then( sub {
+               my ( $req ) = @_;
+
+               log_if_fail "Got /state";
+
+               $req->respond_json( {
+                  pdus => [],
+                  auth_chain => [],
+               } );
+
+               Future->done(1);
+            }),
+            $inbound_server->await_request_get_missing_events( $room_id )
+            ->then( sub {
+               my ( $req ) = @_;
+
+               $req->respond_json( {
+                  events => [],
+               } );
+
+               Future->done(1);
+            }),
+
+            $outbound_client->send_event(
+               event       => $sent_event,
+               destination => $first_home_server,
+            ),
+         );
+      })->then( sub {
+         # creator user should eventually receive the sent event
+         await_event_for( $creator, filter => sub {
+            my ( $event ) = @_;
+            return $event->{type} eq "m.room.message" &&
+                   $event->{event_id} eq $sent_event->{event_id};
+         });
```

Comment:
Might be nice just to put a

    ->then_done(1)

here, causing this final block to yield `1` instead of the actual event it receives.